File: 1.300/232.ImplementQueueUsingStacks.cpp
#include <iostream>
#include <stack>
using namespace std;

//232. 用栈实现队列
class MyQueue {
public:
    /** Initialize your data structure here. */
    MyQueue() {

    }

    /** Push element x to the back of queue. */
    void push(int x) {
        stack<int> tmp;
        while (!empty())
            tmp.push(pop());
        stk.push(x);
        while (!tmp.empty()) {
            stk.push(tmp.top());
            tmp.pop();
        }
    }

    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        int i = stk.top();
        stk.pop();
        return i;
    }

    /** Get the front element. */
    int peek() {
        return stk.top();
    }

    /** Returns whether the queue is empty. */
    bool empty() {
        return stk.empty();
    }
private:
    stack<int> stk;
};

int main(int argc, char const* argv[])
{
    MyQueue cls;
    cls.push(1);
    cls.push(2);
    cout << cls.peek() << " -> 1" << endl;
    cout << cls.pop() << " -> 1" << endl;
    cout << boolalpha << cls.empty() << " -> false" << endl;
    return 0;
}

File: go/1.300/232.ImplementQueueUsingStacks.go
package main

import "fmt"

type MyQueue struct {
	Stack []int
}

/** Initialize your data structure here. */
func Constructor() MyQueue {
	return MyQueue{Stack: []int{}}
}

/** Push element x to the back of queue. */
func (this *MyQueue) Push(x int) {
	this.Stack = append(this.Stack, x)
}

/** Removes the element from in front of queue and returns that element. */
func (this *MyQueue) Pop() int {
	i := this.Stack[0]
	this.Stack = this.Stack[1:]
	return i
}

/** Get the front element. */
func (this *MyQueue) Peek() int {
	if this.Empty() {
		return 0
	}
	return this.Stack[0]
}

/** Returns whether the queue is empty. */
func (this *MyQueue) Empty() bool {
	return len(this.Stack) == 0
}

func main() {
	queue := Constructor()
	queue.Push(1)
	queue.Push(2)
	queue.Push(3)
	queue.Push(4)
	fmt.Println(queue.Pop(), 1)
	queue.Push(5)
	fmt.Println(queue.Pop(), 2)
	fmt.Println(queue.Pop(), 3)
	fmt.Println(queue.Pop(), 4)
	fmt.Println(queue.Pop(), 5)
}
